src/soundloader_wav.cpp

```diff
@@ -0,0 +1,97 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file soundloader_wav.cpp Loading of wav sounds. */
+
+#include "stdafx.h"
+#include "core/bitmath_func.hpp"
+#include "core/math_func.hpp"
+#include "random_access_file_type.h"
+#include "sound_type.h"
+#include "soundloader_type.h"
+
+#include "safeguards.h"
+
+/** Wav file (RIFF/WAVE) sound loader. */
+class SoundLoader_Wav : public SoundLoader {
+public:
+	SoundLoader_Wav() : SoundLoader("wav", "Wav sound loader", 0) {}
+
+	static constexpr uint16_t DEFAULT_SAMPLE_RATE = 11025;
+
+	bool Load(SoundEntry &sound, bool new_format, std::vector<uint8_t> &data) override
+	{
+		RandomAccessFile &file = *sound.file;
+
+		/* Check RIFF/WAVE header. */
+		if (file.ReadDword() != BSWAP32('RIFF')) return false;
+		file.ReadDword(); // Skip data size
+		if (file.ReadDword() != BSWAP32('WAVE')) return false;
+
+		/* Read riff tags */
+		for (;;) {
+			uint32_t tag = file.ReadDword();
+			uint32_t size = file.ReadDword();
+
+			if (tag == BSWAP32('fmt ')) {
+				uint16_t format = file.ReadWord();
+				if (format != 1) return false; // File must be uncompressed PCM
+
+				sound.channels = file.ReadWord();
+				if (sound.channels != 1) return false; // File must be mono.
+
+				sound.rate = file.ReadDword();
+				if (!new_format) sound.rate = DEFAULT_SAMPLE_RATE; // All old samples should be played at 11025 Hz.
+
+				file.ReadDword(); // avg bytes per second
+				file.ReadWord();  // alignment
+
+				sound.bits_per_sample = file.ReadWord();
+				if (sound.bits_per_sample != 8 && sound.bits_per_sample != 16) return false; // File must be 8 or 16 BPS.
+			} else if (tag == BSWAP32('data')) {
+				uint align = sound.channels * sound.bits_per_sample / 8;
+				if (Align(size, align) != size) return false; // Ensure length is aligned correctly for channels and BPS.
+
+				if (size == 0) return true; // No need to continue.
+
+				/* Allocate an extra sample to ensure the runtime resampler doesn't go out of bounds. */
+				data.reserve(size + align);
+				data.resize(size);
+
+				file.ReadBlock(std::data(data), size);
+
+				switch (sound.bits_per_sample) {
+					case 8:
+						/* Convert 8-bit samples from unsigned to signed. */
+						for (auto &sample : data) {
+							sample = sample - 128;
+						}
+						break;
+
+					case 16:
+						/* 16-bit samples in wav files are little endian, and may need to be converted to native endian. */
+						if constexpr (std::endian::native != std::endian::little) {
+							for (auto it = std::begin(data); it != std::end(data); /* nothing */) {
+								std::swap(*it++, *it++);
+							}
+						}
+						break;
+
+					default: NOT_REACHED();
+				}
+
+				return true;
+			} else {
+				break;
```

Comment:
Is there a reason we go from 'Skip unknown chunks' to 'Refuse to decode when there is an unknown chunk'?
If so, wouldn't `return false;` be clearer?
Or is it just a mistake from converting the `switch` to an `if-else`-tree?

Reply:
Ah. This is because there was actually TWO wave file loaders before. The one for Basesets behaves this way as well (and is also if-else), the one for NewGRFs skipped unknown chunks.

So yes, the reason is I took just one of the implementations. I can make it skip unknown chunks.

Reply:
I would go for skipping unknown chunks because:
* I can't find any checks to prevent unknown chunks in NewGRFs, so potentially there now is one that is broken and might be played as raw PCM.
* In the associated catcodec PR the 'validation' for the existence of unknown chunks is removed.
* "The RIFF specification requires that applications ignore chunks they do not recognize and applications may not necessarily use this extra information." I must admit that catcodec only ignored unknown chunks after the data chunk, not before that.

Reply:
Done, and also fixed the case where the `fmt ` chunk is longer than 16 bytes.